prow/github/client.go

```diff
@@ -2598,19 +2598,78 @@ func (c *client) ReopenPR(org, repo string, number int) error {
 // GetRef returns the SHA of the given ref, such as "heads/master".
 //
 // See https://developer.github.com/v3/git/refs/#get-a-reference
+// The gitbub api does prefix matching and might return multiple results,
+// in which case we will return a GetRefTooManyResultsError
 func (c *client) GetRef(org, repo, ref string) (string, error) {
 	durationLogger := c.log("GetRef", org, repo, ref)
 	defer durationLogger()
 
-	var res struct {
-		Object map[string]string `json:"object"`
-	}
+	res := GetRefResponse{}
 	_, err := c.request(&request{
 		method:    http.MethodGet,
 		path:      fmt.Sprintf("/repos/%s/%s/git/refs/%s", org, repo, ref),
 		exitCodes: []int{200},
 	}, &res)
-	return res.Object["sha"], err
+	if err != nil {
+		return "", nil
+	}
+
+	if n := len(res); n > 1 {
+		return "", GetRefTooManyResultsError{org: org, repo: repo, ref: ref, resultsRefs: res.RefNames()}
+	}
+	return res[0].Object.SHA, nil
+}
+
+type GetRefTooManyResultsError struct {
+	org, repo, ref string
+	resultsRefs    []string
+}
+
+func (GetRefTooManyResultsError) Is(err error) bool {
```

Comment:
This signature seems wrong to me. Does this need to be a method of `GetRefTooManyResultsError` rather than a receiverless function? This would have to be called like so: `GetRefTooManyResultsError{}.Is(err)`.

Reply:
This is not meant to be called directly but to suport `errors.Is`: https://golang.org/pkg/errors/#Is

Reply:
Usage looks like this in the end:

```
if errors.Is(err, github.GetRefTooManyResultsError{}) { do something}
```

Reply:
Neat, I didn't know about that.